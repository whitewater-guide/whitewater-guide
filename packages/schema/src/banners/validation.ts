import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { array, bool, mixed, number, object, string } from 'yup';

import type { BannerInput, BannerSourceInput } from '../__generated__/types';
import { BannerKind, BannerPlacement } from '../__generated__/types';

const BannerSourceSchema: ObjectSchema<BannerSourceInput> = object({
  kind: mixed<BannerKind>().oneOf(Object.values(BannerKind)).required(),
  url: string().required(),
}).noUnknown();

export const BannerInputSchema: ObjectSchema<BannerInput> = object({
  id: string().uuid().nullable(),
  slug: string().slug().required(),
  name: string().required(),
  priority: number().integer().required(),
  enabled: bool().required(),
  placement: mixed<BannerPlacement>()
    .oneOf(Object.values(BannerPlacement))
    .required(),
  source: BannerSourceSchema.clone(),
  link: string().url().https().nullable(),
  extras: object().nullable(),
  regions: array(yupSchemas.refInput().required()).required(),
  groups: array(yupSchemas.refInput().required()).required(),
})
  .strict(true)
  .noUnknown(true);
