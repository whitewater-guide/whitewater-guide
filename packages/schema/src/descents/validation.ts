import type { ObjectSchema } from 'yup';
import { bool, date, number, object, string } from 'yup';

import type { DescentInput, DescentLevelInput } from '../__generated__/types';

export const DescentLevelInputSchema: ObjectSchema<DescentLevelInput> = object({
  unit: string().max(20).optional().nullable(),
  value: number().required(),
})
  .strict(true)
  .noUnknown();

export const DescentInputSchema: ObjectSchema<DescentInput> = object({
  id: string().uuid().optional().nullable(),
  sectionId: string().uuid().required().nonNullable(),
  startedAt: date().required(),
  duration: number().integer().min(0).notRequired().nullable(),
  level: DescentLevelInputSchema.clone().notRequired().nullable(),
  comment: string().notRequired().nullable(),
  public: bool().notRequired().nullable(),
})
  .strict(true)
  .noUnknown();
