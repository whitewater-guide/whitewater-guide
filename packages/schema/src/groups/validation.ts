import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { GroupInput } from '../__generated__/types';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema: ObjectSchema<GroupInput> = object({
  id: string().uuid().defined().nullable(),
  name: string().nonEmpty(),
  sku: string()
    .defined()
    .max(255)
    .matches(GROUP_SKU, 'yup:string.sku')
    .nullable(),
})
  .strict(true)
  .noUnknown();
