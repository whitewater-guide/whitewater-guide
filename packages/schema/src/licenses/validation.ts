import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { LicenseInput } from '../__generated__/types';

export const LicenseInputSchema: ObjectSchema<LicenseInput> = object({
  url: string().url().nullable(),
  name: string().min(2).defined().nonNullable(),
  slug: string().slug().nullable(),
})
  .strict(true)
  .noUnknown();
