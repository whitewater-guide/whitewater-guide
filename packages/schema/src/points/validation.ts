import { isArray, isNumber } from 'lodash';
import type { ObjectSchema, Schema } from 'yup';
import { array, object, string } from 'yup';

import type { PointInput } from '../__generated__/types';
import { POITypes } from './POITypes';

export const CoordinateSchema: Schema<CodegenCoordinates> = array()
  .max(3)
  .test({
    name: 'is-coordinate3d-lng',
    test(v) {
      const valid = isArray(v) && isNumber(v[0]) && v[0] >= -180 && v[0] <= 180;
      const path = this.path ? `${this.path}.0` : '0';
      return (
        valid || this.createError({ path, message: 'yup:number.longitude' })
      );
    },
  })
  .test({
    name: 'is-coordinate3d-lat',
    test(v) {
      const valid = isArray(v) && isNumber(v[1]) && v[1] >= -90 && v[1] <= 90;
      const path = this.path ? `${this.path}.1` : '1';
      return (
        valid || this.createError({ path, message: 'yup:number.latitude' })
      );
    },
  })
  .test({
    name: 'is-coordinate3d-alt',
    test(v) {
      const path = this.path ? `${this.path}.2` : '2';
      return (
        (isArray(v) &&
          (v[2] === undefined || v[2] === null || isNumber(v[2]))) ||
        this.createError({ path, message: 'yup:number.altitude' })
      );
    },
  }) as any;

export const PointInputSchema: ObjectSchema<PointInput> = object({
  id: string().uuid().defined().nullable(),
  name: string().nullable(),
  description: string().nullable(),
  coordinates: CoordinateSchema.clone(),
  kind: string().defined().oneOf(POITypes),
})
  .strict(true)
  .noUnknown();
