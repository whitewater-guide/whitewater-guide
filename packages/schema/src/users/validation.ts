import type { ObjectSchema } from 'yup';
import { bool, object, string } from 'yup';

import type { EditorSettingsInput, UserInput } from '../__generated__/types';
import { LANGUAGES } from '../i18n';

export const UserInputSchema: ObjectSchema<UserInput> = object({
  name: string().notRequired().min(1),
  avatar: string().notRequired().nullable(),
  language: string().oneOf(LANGUAGES).optional(),
  imperial: bool().notRequired(),
  email: string().email().notRequired().nullable(),
})
  .strict(true)
  .noUnknown();

export const EditorSettingsSchema: ObjectSchema<EditorSettingsInput> = object({
  language: string().oneOf(LANGUAGES).required(),
})
  .strict(true)
  .noUnknown();
