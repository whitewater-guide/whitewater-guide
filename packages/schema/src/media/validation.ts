import type { ObjectSchema } from 'yup';
import { mixed, number, object, string, tuple } from 'yup';

import type { MediaInput } from '../__generated__/types';
import { MediaKind } from '../__generated__/types';
import { LicenseInputSchema } from '../licenses';

export const MediaInputSchema: ObjectSchema<MediaInput> = object({
  id: string().uuid().nullable(),
  description: string().nullable(),
  copyright: string().nullable(),
  license: LicenseInputSchema.clone().nullable(),
  url: string().nonEmpty(),
  kind: mixed<MediaKind>().defined().oneOf(Object.values(MediaKind)),
  resolution: tuple([
    number().required().integer().positive(),
    number().required().integer().positive(),
  ]).notRequired(),
  weight: number().integer().nullable(),
})
  .strict(true)
  .noUnknown();
