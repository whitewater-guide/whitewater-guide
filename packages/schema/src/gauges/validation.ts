import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { mixed, object, string } from 'yup';

import type { GaugeInput } from '../__generated__/types';
import { PointInputSchema } from '../points';

export const GaugeInputSchema: ObjectSchema<GaugeInput> = object({
  id: string().uuid().nullable(),
  name: string().nonEmpty(),
  code: string().nonEmpty(),
  levelUnit: string().nullable(),
  flowUnit: string().nullable(),
  timezone: string().nullable(),
  location: PointInputSchema.clone().nullable(),
  requestParams: mixed().nullable(),
  url: string().url().nullable(),
  source: yupSchemas.refInput().required(),
})
  .strict(true)
  .noUnknown();
