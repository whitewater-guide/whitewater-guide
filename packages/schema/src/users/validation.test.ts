import { createSafeValidator } from '@whitewater-guide/validation';

import type { UserInput } from '../__generated__/types';
import { UserInputSchema } from './validation';

const validator = createSafeValidator(UserInputSchema);

type TestValue = [string, UserInput];
type IncorrectTestValue = [string, any];

const correctValues: TestValue[] = [
  [
    'full value',
    {
      name: 'foo',
      avatar: 'bar',
      imperial: true,
      language: 'en',
      email: 'foo@bar.com',
    },
  ],
  ['empty avatar', { name: 'foo', avatar: '', imperial: true, language: 'en' }],
  [
    'null avatar',
    { name: 'foo', avatar: null, imperial: true, language: 'en' },
  ],
  ['no name', { avatar: 'bar', imperial: true, language: 'en' }],
  ['no avatar', { name: 'foo', imperial: true, language: 'en' }],
  ['no imperial', { name: 'foo', avatar: 'bar', language: 'en' }],
  ['no language', { name: 'foo', avatar: 'bar', imperial: true }],
  ['no email', { name: 'foo', avatar: 'bar', imperial: true, language: 'en' }],
  [
    'null email',
    { name: 'foo', avatar: 'bar', imperial: true, language: 'en', email: null },
  ],
];

const incorrectValues: IncorrectTestValue[] = [
  ['empty name', { name: '' }],
  ['bad email', { language: 'en', email: 'bar' }],
  ['extra fields', { language: 'en', foo: 'bar' }],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  const error = validator(value);
  expect(error).not.toBeNull();
  expect(error).toMatchSnapshot();
});
