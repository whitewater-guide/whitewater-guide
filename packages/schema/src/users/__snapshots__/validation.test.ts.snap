// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad email 1`] = `
{
  "email": {
    "key": "yup:string.email",
    "options": {
      "path": "email",
    },
  },
}
`;

exports[`should be invalid for empty name 1`] = `
{
  "name": {
    "key": "yup:string.min",
    "options": {
      "min": 1,
      "path": "name",
    },
  },
}
`;

exports[`should be invalid for extra fields 1`] = `
{
  "": {
    "key": "yup:object.noUnknown",
    "options": {
      "path": "this",
    },
  },
}
`;
